firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/TesterSignInManager.java

```diff
@@ -104,50 +107,68 @@ void onActivityStarted(Activity activity) {
       return;
     } else {
       // Throw error if app reentered during sign in
-      if (this.isCurrentlySigningIn()) {
-        LogWrapper.getInstance().e("App Resumed without sign in flow completing.");
-        this.setCanceledAuthenticationError();
+      synchronized (signInTaskLock) {
+        if (awaitingResultFromBrowser()) {
+          LogWrapper.getInstance().e("App Resumed without sign in flow completing.");
+          setSignInTaskCompletionError(
+              new FirebaseAppDistributionException(
+                  Constants.ErrorMessages.AUTHENTICATION_CANCELED, AUTHENTICATION_CANCELED));
+        }
       }
     }
   }
 
   @NonNull
   public Task<Void> signInTester() {
-
     if (signInStorage.getSignInStatus()) {
       LogWrapper.getInstance().v(TAG + "Tester is already signed in.");
       return Tasks.forResult(null);
     }
 
     synchronized (signInTaskLock) {
-      if (this.isCurrentlySigningIn()) {
+      if (signInTaskCompletionSource != null
+          && !signInTaskCompletionSource.getTask().isComplete()) {
         LogWrapper.getInstance()
             .v(TAG + "Detected In-Progress sign in task. Returning the same task.");
         return signInTaskCompletionSource.getTask();
       }
 
       signInTaskCompletionSource = new TaskCompletionSource<>();
+      hasBeenSentToBrowserForCurrentTask = false;
 
       firebaseInstallationsApiProvider
           .get()
           .getId()
-          .addOnSuccessListener(getFidGenerationOnSuccessListener())
           .addOnFailureListener(
-              e -> {
-                LogWrapper.getInstance().e(TAG + "Fid retrieval failed.", e);
-                setSignInTaskCompletionError(
-                    new FirebaseAppDistributionException(
-                        ErrorMessages.AUTHENTICATION_ERROR, AUTHENTICATION_FAILURE, e));
-              });
+              handleTaskFailure(ErrorMessages.AUTHENTICATION_ERROR, Status.AUTHENTICATION_FAILURE))
+          .onSuccessTask(combineWithResultOf(() -> lifecycleNotifier.getForegroundActivity()))
```

Comment:
Will this be ok in the case where we get the activity but we don't have the lock and once we get the lock the activity changes? 

Foreground activity with no lock-> Lock granted -> Different foreground activity is now there but we continue now that we have the lock.

Reply:
Oh great catch! That would break. I'll see what I can do about that.

Reply:
I made a change in the success listener below, and left some clarifying comments. I think that helps but let me know what you think.